frame/system/benchmarking/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+
+use codec::Encode;
+use sp_std::vec;
+use sp_std::prelude::*;
+use sp_core::{ChangesTrieConfiguration, storage::well_known_keys};
+use sp_runtime::traits::Hash;
+use frame_benchmarking::{benchmarks, account};
+use frame_support::storage::{self, StorageMap};
+use frame_system::{Module as System, Call, RawOrigin, DigestItemOf, AccountInfo};
+
+mod mock;
+
+const SEED: u32 = 0;
+
+pub struct Module<T: Trait>(System<T>);
+pub trait Trait: frame_system::Trait {}
+
+benchmarks! {
+	_ { }
+
+	remark {
+		// # of Bytes
+		let b in 0 .. 16_384;
+		let remark_message = vec![1; b as usize];
+		let caller = account("caller", 0, SEED);
+	}: _(RawOrigin::Signed(caller), remark_message)
+
+	set_heap_pages {
+		// Heap page size
+		let i in 0 .. u32::max_value();
+	}: _(RawOrigin::Root, i.into())
+
+	// `set_code` was not benchmarked because it is pretty hard to come up with a real
+	// Wasm runtime to test the upgrade with. But this is okay because we will make
+	// `set_code` take a full block anyway.
+
+	set_code_without_checks {
+		// Version number
+		let b in 0 .. 16_384;
+		let code = vec![1; b as usize];
+	}: _(RawOrigin::Root, code)
+	verify {
+		let current_code = storage::unhashed::get_raw(well_known_keys::CODE).ok_or("Code not stored.")?;
+		assert_eq!(current_code.len(), b as usize);
+	}
+
+	set_changes_trie_config {
+		let d in 0 .. 1000;
+
+		let digest_item = DigestItemOf::<T>::Other(vec![]);
+
+		for i in 0 .. d {
+			System::<T>::deposit_log(digest_item.clone());
+		}
+		let changes_trie_config = ChangesTrieConfiguration {
+			digest_interval: d,
+			digest_levels: d,
+		};
+	}: _(RawOrigin::Root, Some(changes_trie_config))
+	verify {
+		assert_eq!(System::<T>::digest().logs.len(), (d + 1) as usize)
+	}
+
+	set_storage {
+		let i in 1 .. 1000;
+
+		// Set up i items to add
+		let mut items = Vec::new();
+		for j in 0 .. i {
+			let hash = (i, j).using_encoded(T::Hashing::hash).as_ref().to_vec();
+			items.push((hash.clone(), hash.clone()));
+		}
+	}: _(RawOrigin::Root, items)
+	verify {
+		let last_hash = (i, i - 1).using_encoded(T::Hashing::hash);
+		let value = storage::unhashed::get_raw(last_hash.as_ref()).ok_or("No value stored")?;
+		assert_eq!(value, last_hash.as_ref().to_vec());
+	}
+
+	kill_storage {
+		let i in 1 .. 1000;
+
+		// Add i items to storage
+		let mut items = Vec::new();
+		for j in 0 .. i {
+			let hash = (i, j).using_encoded(T::Hashing::hash).as_ref().to_vec();
+			storage::unhashed::put_raw(&hash, &hash);
+			items.push(hash);
+		}
+
+		// We will verify this value is removed
+		let last_hash = (i, i - 1).using_encoded(T::Hashing::hash);
+		let value = storage::unhashed::get_raw(last_hash.as_ref()).ok_or("No value stored")?;
+		assert_eq!(value, last_hash.as_ref().to_vec());
+
+	}: _(RawOrigin::Root, items)
+	verify {
+		assert_eq!(storage::unhashed::get_raw(last_hash.as_ref()), None);
+	}
+
+	kill_prefix {
+		let p in 1 .. 1000;
+
+		let prefix = p.using_encoded(T::Hashing::hash).as_ref().to_vec();
```

Comment:
This prefix is always the same length, correct?
Can we assume this about the input `Key` as well? (The type alias just says `Vec<u8>`)

Reply:
Prefix length and key size shouldn't matter. Just number of prefixes